Readme.md
# retext-soundex [![Build Status](https://img.shields.io/travis/wooorm/retext-soundex.svg?style=flat)](https://travis-ci.org/wooorm/retext-soundex) [![Coverage Status](https://img.shields.io/coveralls/wooorm/retext-soundex.svg?style=flat)](https://coveralls.io/r/wooorm/retext-soundex?branch=master)

**[Retext](https://github.com/wooorm/retext)** implementation of the [Soundex](http://en.wikipedia.org/wiki/Soundex) algorithm.

## Installation

[npm](https://docs.npmjs.com/cli/install):

```bash
$ npm install retext-soundex
```

[Component.js](https://github.com/componentjs/component):

```bash
$ component install wooorm/retext-soundex
```

[Bower](http://bower.io/#install-packages):

```bash
$ bower install retext-soundex
```

[Duo](http://duojs.org/#getting-started):

```javascript
var soundex = require('wooorm/retext-soundex');
```

## Usage

```javascript
var Retext = require('retext');
var visit = require('retext-visit');
var inspect = require('retext-inspect');
var soundex = require('retext-soundex');

var retext = new Retext()
    .use(inspect)
    .use(visit)
    .use(soundex);

retext.parse('A simple english sentence.', function (err, tree) {
    tree.visit(tree.WORD_NODE, function (node) {
        console.log(node);
    });
    /*
     * WordNode[1] [data={"phonetics":"A000"}]
     * └─ TextNode: 'A'
     * WordNode[1] [data={"phonetics":"S514"}]
     * └─ TextNode: 'simple'
     * WordNode[1] [data={"phonetics":"E5242"}]
     * └─ TextNode: 'english'
     * WordNode[1] [data={"phonetics":"S5352"}]
     * └─ TextNode: 'sentence'
     */
});
```

You can also combine it with a stemmer (such as [retext-porter-stemmer](https://github.com/wooorm/retext-porter-stemmer) or [retext-lancaster-stemmer](https://github.com/wooorm/retext-lancaster-stemmer)).

```javascript
var Retext = require('retext');
var visit = require('retext-visit');
var inspect = require('retext-inspect');
var soundex = require('retext-soundex');
var stemmer = require('retext-porter-stemmer');

var retext = new Retext()
    .use(inspect)
    .use(visit)
    .use(soundex)
    /* make sure to attach the stemmer after soundex. */
    .use(stemmer);

retext.parse('A detestable paragraph.', function (err, tree) {
    tree.visit(tree.WORD_NODE, function (node) {
        console.log(node);
    });
    /*
     * WordNode[1] [data={"stem":"a","phonetics":"A000","stemmedPhonetics":"A000"}]
     * └─ TextNode: 'A'
     * WordNode[1] [data={"stem":"detest","phonetics":"D32314","stemmedPhonetics":"D323"}]
     * └─ TextNode: 'detestable'
     * WordNode[1] [data={"stem":"paragraph","phonetics":"P6261","stemmedPhonetics":"P6261"}]
     * └─ TextNode: 'paragraph'
     */
});
```

## API

None, **retext-soundex** automatically detects the phonetics of each [`WordNode`](https://github.com/wooorm/textom#textomwordnode-nlcstwordnode) (using **[wooorm/soundex-code](https://github.com/wooorm/soundex-code)**), and stores the phonetics in `node.data.phonetics`. If a stemmer is used, the stemmed phonetics are stored in `node.data.stemmedPhonetics`.

## Performance

On a MacBook Air, **retext** performs about 13% slower with **retext-soundex**.

```text
           retext w/o retext-soundex
  225 op/s » A paragraph (5 sentences, 100 words)
   25 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)

           retext w/ retext-soundex
  196 op/s » A paragraph (5 sentences, 100 words)
   18 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)
```

## Related

- [retext-metaphone](https://github.com/wooorm/retext-metaphone) — The original Metaphone algorithm;
- [retext-double-metaphone](https://github.com/wooorm/retext-double-metaphone) — The Double Metaphone algorithm;

## License

[MIT](LICENSE) © [Titus Wormer](http://wooorm.com)


History.md

0.1.6 / 2015-01-19
==================

  * Update copyright notice in `LICENSE` to include 2015
  * Add link to whole license in `Readme.md`
  * Add Duo as an instalation method in `Readme.md`
  * Add links to installation methods in `Readme.md`
  * Refactor fences code blocks in `Readme.md`
  * Refactor to adhere to strict jsdoc style
  * Add jscs-jsdoc configuration to `.jscs.json`
  * Add jscs-jsdoc as a dev-dependency
  * Refactor npm scripts for changes in npm
  * Update eslint
  * Refactor to use generic `changeinside` in new retext
  * Update retext to stable release

0.1.5 / 2014-12-14
==================

 * Add benchmark
 * Fix wording in `Readme.md`
 * Add link to personal website to copyright in `Readme.md`
 * Update eslint, retext

0.1.4 / 2014-11-22
==================

 * Refactor `Readme.md`
 * Add flat badges to `Readme.md`
 * Fix incorrect executive rights on `test.js`
 * Refactor npm script targets in `package.json`
 * Add retext-inspect to dev-dependecies
 * Update retext, soundex-code

0.1.3 / 2014-10-28
==================

 * Add `.eslintrc`
 * Refactor to disallow space after object keys
 * Refactor for changes in retext
 * Update soundex-code, eslint, mocha, retext, retext-content, retext-porter-stemmer, retext-visit

0.1.2 / 2014-10-21
==================

 * Refactor .jscs.json
 * Update retext

0.1.1 / 2014-10-15
==================

 * Update tests for changes in retet-visit
 * Update retext, retext-content, retext-porter-stemmer, retext-visit
 * Update .gitignore, .npmignore, bower ignore
 * Move spec to test.js

0.1.0 / 2014-09-27
==================

 * Remove incorrect history
 * Refactor Readme.md
 * Remove example
 * Refactor API
 * Remove browser test
 * Update .gitignore, .npmignore
 * Fix property order in component.json, package.json, bower.json
 * Update copyright in Readme.md
 * Remove testling
 * Refactor spec
 * Update dependencies


test.js
'use strict';

/*
 * Dependencies.
 */

var soundex,
    Retext,
    inspect,
    stemmer,
    visit,
    content,
    assert;

soundex = require('./');
Retext = require('retext');
inspect = require('retext-inspect');
stemmer = require('retext-porter-stemmer');
visit = require('retext-visit');
content = require('retext-content');
assert = require('assert');

/*
 * Retext.
 */

var retext,
    retextWithStemmer;

retext = new Retext()
    .use(inspect)
    .use(content)
    .use(visit)
    .use(soundex);

retextWithStemmer = new Retext()
    .use(inspect)
    .use(content)
    .use(visit)
    .use(soundex)
    .use(stemmer);

/*
 * Fixtures.
 */

var sentence,
    otherWords,
    otherPhonetics,
    stemmedOtherPhonetics;

sentence = 'A simple, english, sentence';
otherWords = ['A', 'detestable', 'vile', 'paragraph'];
otherPhonetics = ['A000', 'D32314', 'V400', 'P6261'];
stemmedOtherPhonetics = ['A000', 'D323', 'V400', 'P6261'];

/*
 * Tests.
 */

describe('soundex()', function () {
    it('should be of type `function`', function () {
        assert(typeof soundex === 'function');
    });

    retext.parse(sentence, function (err, tree) {
        it('should not throw', function (done) {
            done(err);
        });

        it('should process each `WordNode`', function () {
            tree.visit(tree.WORD_NODE, function (wordNode) {
                assert('phonetics' in wordNode.data);
            });
        });

        it('should set `phonetics` to `null` when `WordNode` (no ' +
            'longer?) has a value',
            function () {
                tree.visit(tree.WORD_NODE, function (wordNode) {
                    wordNode.removeContent();

                    assert(wordNode.data.phonetics === null);
                });
            }
        );

        it('should re-process `WordNode`s when their values change',
            function () {
                var index;

                index = -1;

                tree.visit(tree.WORD_NODE, function (wordNode) {
                    index++;

                    wordNode.replaceContent(otherWords[index]);

                    assert(wordNode.data.phonetics === otherPhonetics[index]);
                });
            }
        );
    });
});

describe('soundex() with a stemmer', function () {
    retextWithStemmer.parse(sentence, function (err, tree) {
        it('should not throw', function (done) {
            done(err);
        });

        it('should process `stem` in each `WordNode`', function () {
            tree.visit(tree.WORD_NODE, function (wordNode) {
                assert('stemmedPhonetics' in wordNode.data);
            });
        });

        it('should set `stemmedPhonetics` to `null` when `WordNode` (no ' +
            'longer?) has a value',
            function () {
                tree.visit(tree.WORD_NODE, function (wordNode) {
                    wordNode.removeContent();

                    assert(wordNode.data.stemmedPhonetics === null);
                });
            }
        );

        it('should re-process `WordNode`s when their stem changes',
            function () {
                var index;

                index = -1;

                tree.visit(tree.WORD_NODE, function (wordNode) {
                    index++;

                    wordNode.replaceContent(otherWords[index]);

                    assert(
                        wordNode.data.stemmedPhonetics ===
                        stemmedOtherPhonetics[index]
                    );
                });
            }
        );
    });
});


spec/retext-soundex.spec.js
'use strict';

/**
 * Dependencies.
 */

var soundex,
    stemmer,
    visit,
    content,
    Retext,
    assert;

soundex = require('..');
Retext = require('retext');
visit = require('retext-visit');
content = require('retext-content');
stemmer = require('retext-porter-stemmer');
assert = require('assert');

/**
 * Retext.
 */

var retext,
    retextWithStemmer;

retext = new Retext()
    .use(content)
    .use(visit)
    .use(soundex);

retextWithStemmer = new Retext()
    .use(content)
    .use(visit)
    .use(soundex)
    .use(stemmer);

/**
 * Fixtures.
 */

var sentence,
    otherWords,
    otherPhonetics,
    stemmedOtherPhonetics;

sentence = 'A simple, english, sentence';
otherWords = ['A', 'detestable', 'vile', 'paragraph'];
otherPhonetics = ['A000', 'D32314', 'V400', 'P6261'];
stemmedOtherPhonetics = ['A000', 'D323', 'V400', 'P6261'];

/**
 * Tests.
 */

describe('soundex()', function () {
    it('should be of type `function`', function () {
        assert(typeof soundex === 'function');
    });

    retext.parse(sentence, function (err, tree) {
        it('should not throw', function (done) {
            done(err);
        });

        it('should process each `WordNode`', function () {
            tree.visitType(tree.WORD_NODE, function (wordNode) {
                assert('phonetics' in wordNode.data);
            });
        });

        it('should set `phonetics` to `null` when `WordNode` (no ' +
            'longer?) has a value',
            function () {
                tree.visitType(tree.WORD_NODE, function (wordNode) {
                    wordNode.removeContent();

                    assert(wordNode.data.phonetics === null);
                });
            }
        );

        it('should re-process `WordNode`s when their values change',
            function () {
                var index;

                index = -1;

                tree.visitType(tree.WORD_NODE, function (wordNode) {
                    index++;

                    wordNode.replaceContent(otherWords[index]);

                    assert(wordNode.data.phonetics === otherPhonetics[index]);
                });
            }
        );
    });
});

describe('soundex() with a stemmer', function () {
    retextWithStemmer.parse(sentence, function (err, tree) {
        it('should not throw', function (done) {
            done(err);
        });

        it('should process `stem` in each `WordNode`', function () {
            tree.visitType(tree.WORD_NODE, function (wordNode) {
                assert('stemmedPhonetics' in wordNode.data);
            });
        });

        it('should set `stemmedPhonetics` to `null` when `WordNode` (no ' +
            'longer?) has a value',
            function () {
                tree.visitType(tree.WORD_NODE, function (wordNode) {
                    wordNode.removeContent();

                    assert(wordNode.data.stemmedPhonetics === null);
                });
            }
        );

        it('should re-process `WordNode`s when their stem changes',
            function () {
                var index;

                index = -1;

                tree.visitType(tree.WORD_NODE, function (wordNode) {
                    index++;

                    wordNode.replaceContent(otherWords[index]);

                    assert(
                        wordNode.data.stemmedPhonetics ===
                        stemmedOtherPhonetics[index]
                    );
                });
            }
        );
    });
});


index.js
/**
 * Dependencies.
 */

var Retext = require('wooorm/retext@0.5.0');
var soundex = require('wooorm/retext-soundex@0.1.6');
var stemmer = require('wooorm/retext-porter-stemmer@0.2.5');
var dom = require('wooorm/retext-dom@0.3.2');
var visit = require('wooorm/retext-visit@0.2.6');

/**
 * Retext.
 */

var retext = new Retext()
    .use(visit)
    .use(dom)
    .use(soundex)
    .use(stemmer);

/**
 * DOM elements.
 */

var $input = document.getElementsByTagName('textarea')[0];
var $output = document.getElementsByTagName('div')[0];
var $stem = document.getElementsByTagName('input')[0];

/**
 * Get a color representation from a string.
 */

var style = document.styleSheets[0];

var phonetics = {};

function hashCode(str) {
    var hash = 0;
    for (var i = 0; i < str.length; i++) {
       hash = str.charCodeAt(i) + ((hash << 5) - hash);
    }
    return hash;
}

function intToARGB(i){
    return ((i>>24)&0xFF).toString(16) +
           ((i>>16)&0xFF).toString(16) +
           ((i>>8)&0xFF).toString(16) +
           (i&0xFF).toString(16);
}

function getColorFromString(value) {
    value = intToARGB(hashCode(value)).slice(0, 6);

    while (value.length < 6) {
        value = '0' + value;
    }

    return '#' + value;
}

/**
 * Add a CSS rule.
 */

function addCSSRule(sheet, selector, rules) {
    if(sheet.insertRule) {
        sheet.insertRule(selector + '{' + rules + '}');
    } else {
        sheet.addRule(selector, rules);
    }
}

/**
 * Callback when new phonetics are calculated.
 */

function onphonetics(phonetic) {
    var color;

    if (phonetic in phonetics) {
        return;
    }

    color = getColorFromString(phonetic)
    phonetics[phonetic] = color;

    addCSSRule(style, '[title="' + phonetic + '"]', 'color:' + color);
}

/**
 * Events
 */

var shouldUseStemmedPhonetics = $stem.checked;

function onshouldstemchange() {
    shouldUseStemmedPhonetics = $stem.checked;

    oninputchange();
}

var tree;

function oninputchange() {
    if (tree) {
        tree.toDOMNode().parentNode.removeChild(tree.toDOMNode());
    }

    retext.parse($input.value, function (err, root) {
        if (err) throw err;

        tree = root;

        tree.visit(function (node) {
            if (!node.DOMTagName || !node.data.phonetics) {
                return;
            }

            phonetic = node.data.phonetics;

            if (shouldUseStemmedPhonetics) {
                phonetic = node.data.stemmedPhonetics;
            }

            onphonetics(phonetic);

            node.toDOMNode().setAttribute('title', phonetic);
        });

        $output.appendChild(tree.toDOMNode());
    });
}

$input.addEventListener('input', oninputchange);
$stem.addEventListener('change', onshouldstemchange);

onshouldstemchange();


benchmark.js
'use strict';

var Retext,
    soundex;

/*
 * Dependencies.
 */

Retext = require('retext');
soundex = require('./');

/*
 * Dependencies.
 */

var retext,
    retextWithSoundex;

retext = new Retext();
retextWithSoundex = new Retext().use(soundex);

/*
 * Test data: A (big?) article (w/ 100 paragraphs, 500
 * sentences, 10,000 words);
 *
 * Source:
 *   http://www.gutenberg.org/files/10745/10745-h/10745-h.htm
 */

var sentence,
    paragraph,
    section;

sentence = 'Where she had stood was clear, and she was gone since Sir ' +
    'Kay does not choose to assume my quarrel.';

paragraph = 'Thou art a churlish knight to so affront a lady ' +
    'he could not sit upon his horse any longer. ' +
    'For methinks something hath befallen my lord and that he ' +
    'then, after a while, he cried out in great voice. ' +
    'For that light in the sky lieth in the south ' +
    'then Queen Helen fell down in a swoon, and lay. ' +
    'Touch me not, for I am not mortal, but Fay ' +
    'so the Lady of the Lake vanished away, everything behind. ' +
    sentence;

section = paragraph + Array(10).join('\n\n' + paragraph);

suite('retext w/o retext-soundex', function () {
    bench('A paragraph (5 sentences, 100 words)', function (done) {
        retext.parse(paragraph, done);
    });

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function (done) {
            retext.parse(section, done);
        }
    );
});

suite('retext w/ retext-soundex', function () {
    bench('A paragraph (5 sentences, 100 words)', function (done) {
        retextWithSoundex.parse(paragraph, done);
    });

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function (done) {
            retextWithSoundex.parse(section, done);
        }
    );
});
